reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt

```diff
@@ -0,0 +1,270 @@
+
+            if (processedFindings.isNotEmpty()) {
+                if (licenseFileFindings.isNotEmpty()) {
+                    add { LICENSE_SEPARATOR }
+                }
+
+                add { "The following copyrights and licenses were found in the source code of this package:\n\n" }
+            }
+
+            addProcessedFindings(processedFindings)
+
+            if (licenseFileFindings.isEmpty() && processedFindings.isEmpty()) {
+                log.error { "No license information was added for package ${id.toCoordinates()}." }
+            }
+        }
+    }
+
+    private fun MutableList<() -> String>.addProcessedFindings(processedFindings: LicenseFindingsMap) {
+        processedFindings.entries.forEachIndexed { index, (license, copyrights) ->
+            input.licenseTextProvider.getLicenseText(license)?.let { licenseText ->
+                copyrights.forEach { copyright ->
+                    add { "$copyright\n" }
+                }
+                if (copyrights.isNotEmpty()) {
+                    add { "\n" }
+                }
+
+                add { licenseText }
+
+                if (index < processedFindings.size - 1) {
+                    add { LICENSE_SEPARATOR }
+                }
+            }
+        }
+    }
+
+    private fun MutableList<() -> String>.addLicenseFileFindings(
+        licenseFileFindings: Map<String, LicenseFindingsMap>,
+        archiveDir: File
+    ) {
+        licenseFileFindings.forEach { (file, findings) ->
+            add { "This package contains the file $file with the following contents:\n\n" }
+            add { "${archiveDir.resolve(file).readText()}\n" }
+            val allCopyrights = findings.values.flatten().toSet()
+            val processedCopyrights = CopyrightStatementsProcessor().process(allCopyrights).toMutableSet()
+            if (processedCopyrights.isNotEmpty()) {
+                add { "The following copyright holder information relates to the license(s) above:\n\n" }
+            }
+            processedCopyrights.forEach { copyright ->
+                add { "$copyright\n" }
+            }
+        }
+    }
+
+    private fun getFindingsForLicenseFiles(
+        scanResult: ScanResult,
+        archiveDir: File,
+        licenseFindingsMap: LicenseFindingsMap
+    ): MutableMap<String, LicenseFindingsMap> {
+        val matcher = FileMatcher(LICENSE_FILENAMES)
+        val licenseFiles = mutableMapOf<String, LicenseFindingsMap>()
+
+        archiveDir.walkTopDown().forEach { file ->
+            val relativePath = archiveDir.toPath().relativize(file.toPath())
+            if (matcher.matches(relativePath.toString())) {
+                val licenses = findLicensesForFile(scanResult, relativePath)
+                val copyrightsInFile = findCopyrightsForFile(scanResult, relativePath)
+
+                val licensesWithCopyrights = licenses.associateWith { license ->
+                    // Remove copyrights which are already contained in the license file.
```

Comment:
optional: I'd find it more readable when the body of the associateWith was extracted to a function.

Reply:
I have moved a slightly larger code block into a separate function.